### DISCUSSION TITLE
Uploading file via graphql

### DISCUSSION THREAD

**Discussion by riteshnoronha** [Orange-OpenSource/hurl#3220] [General] [answered]
Hi Everyone,

Thanks for this amazing tool. I'm currently trying to use hurl to upload files via graphql. I just can't seem to get it to work. Wondering if anyone can help me out. 

I have tried various versions for file upload syntax 
- `< file,basic.json;`
- `@file,basic.json;`
-  `file,basic.json;`

But nothing seems to work. The server reports empty content error, wondering if anyone has insights on how to address this. 

POST {{GRAPHQL_EP}}
Authorization: Bearer {{API_KEY}}
Content-Type: multipart/form-data; boundary=------WebKitFormBoundaryyVcJ5VwZUjJvL162

```
------WebKitFormBoundaryyVcJ5VwZUjJvL162
Content-Disposition: form-data; name="operations"
{
  "operationName": "Upload",
  "query": "mutation Upload($doc: Upload!, $projectId: ID!) {  upload(input: {doc: $doc, projectId: $projectId}) { errors __typename } }",
  "variables": {
    "doc": null,
    "projectId": "{{projectId}}"
  }
}
------WebKitFormBoundaryyVcJ5VwZUjJvL162
Content-Disposition: form-data; name="map"

{ "1": ["variables.doc"] }

------WebKitFormBoundaryyVcJ5VwZUjJvL162
Content-Disposition: form-data; name="1"; filename="basic.json"
Content-Type: application/json

file,basic.json;
------WebKitFormBoundaryyVcJ5VwZUjJvL162--
```

HTTP 200

**Comment by jcamiel** [comment]
Hi @riteshnoronha 

When using multiparty form data as a text body, we don't support sending external files (see https://hurl.dev/docs/request.html#multipart-form-data). 

You can try to use a `[MultipartFormData]` section to see if you get better result. Passing the GraphQL body value might be tricky though, I will check it when I come back to my computer.

**Reply by riteshnoronha** [reply]
I did give that a shot but did not get the results, 
```
[MultipartFormData]
operations: {"query": "mutation Upload($doc: Upload!, $projectId: ID!) { Upload(input: {doc: $doc, projectId: $projectId}) { errors } }", "variables": { "doc": null,  "projectId": "{{projectId}}"}}
map: { "1": ["variables.doc"] }
file: file,basic.json;type=application/json

```
i was trying to mimic the graphql file upload spec 
https://github.com/jaydenseric/graphql-multipart-request-spec

**Reply by riteshnoronha** [reply]
Also i was wondering, is there a passthru mechanism to curl directly?

**Reply by jcamiel** [reply]
Hi @riteshnoronha, there is no pass through, Hurl is a binary using libcurl, so the application is calling libcurl APIs through the file format.

**Reply by riteshnoronha** [reply, marked as answer]
Found the solution. 

```
POST {{GRAPHQL_EP}}
Authorization: Bearer {{API_KEY}}
Content-Type: multipart/form-data
[Options]
very-verbose: true

[MultipartFormData]
operations: {"query": "mutation Upload($doc: Upload!, $projectId: ID!) { Upload(input: {doc: $doc, projectId: $projectId}) { errors } }", "variables": { "doc": null,  "projectId": "{{projectId}}"}}
map: { "0": ["variables.doc"] }
0: file,basic.json;type=application/json

HTTP 200
[Asserts]
jsonpath "$.data.upload.errors" isEmpty
```

Just in case anyone is interested.
